src/payment/strategies/paypal-commerce/paypal-commerce-payment-strategy.ts

```diff
@@ -1,47 +1,88 @@
+import { Cart } from '../../../cart';
 import { CheckoutStore, InternalCheckoutSelectors } from '../../../checkout';
+import { InvalidArgumentError, MissingDataError, MissingDataErrorType } from '../../../common/error/errors';
 import { OrderActionCreator, OrderRequestBody } from '../../../order';
 import { OrderFinalizationNotRequiredError } from '../../../order/errors';
 import { PaymentArgumentInvalidError } from '../../errors';
 import PaymentActionCreator from '../../payment-action-creator';
+import PaymentMethod from '../../payment-method';
+import PaymentMethodActionCreator from '../../payment-method-action-creator';
 import { PaymentInitializeOptions, PaymentRequestOptions } from '../../payment-request-options';
+import PaymentStrategyActionCreator from '../../payment-strategy-action-creator';
 import PaymentStrategy from '../payment-strategy';
 
-import { PaypalCommercePaymentProcessor, PaypalCommerceRequestSender } from './index';
+import { ApproveDataOptions, ButtonsOptions, PaypalCommerceInitializationData, PaypalCommercePaymentProcessor, PaypalCommerceScriptOptions } from './index';
 
 export default class PaypalCommercePaymentStrategy implements PaymentStrategy {
+    private _orderId?: string;
+    private _paymentMethod?: PaymentMethod;
+    private _isPaypalButton: boolean = false;
 
     constructor(
         private _store: CheckoutStore,
         private _orderActionCreator: OrderActionCreator,
         private _paymentActionCreator: PaymentActionCreator,
-        private _paypalCommerceRequestSender: PaypalCommerceRequestSender,
-        private _paypalCommercePaymentProcessor: PaypalCommercePaymentProcessor
+        private _paymentMethodActionCreator: PaymentMethodActionCreator,
+        private _paymentStrategyActionCreator: PaymentStrategyActionCreator,
+        private _paypalCommercePaymentProcessor: PaypalCommercePaymentProcessor,
+        private _credit: boolean = false
     ) {}
 
-    initialize({ paypalcommerce }: PaymentInitializeOptions): Promise<InternalCheckoutSelectors> {
-        this._paypalCommercePaymentProcessor.initialize({ overlay: paypalcommerce && paypalcommerce.overlay });
+    async initialize({ methodId, paypalcommerce }: PaymentInitializeOptions): Promise<InternalCheckoutSelectors> {
+        const state = await this._store.dispatch(this._paymentMethodActionCreator.loadPaymentMethod(methodId));
+        this._paymentMethod = state.paymentMethods.getPaymentMethodOrThrow(methodId);
+        const { initializationData } = this._paymentMethod;
 
-        return Promise.resolve(this._store.getState());
+        if (initializationData.orderId) {
+            this._orderId = initializationData.orderId;
```

Comment:
It seems you're not rendering the PayPal button if the shopper has already started the flow elsewhere (i.e.: cart page). How would the caller know when they need to render a normal submit button vs not render one because PayPal will insert one themselves? Is it through `isShowEmbeddedSubmitButton` getter you've added? If that's the case, I don't particularly like the way we reveal that information through `CheckoutState`, because it is a provider-specific piece of information that is surfaced through an interface that's not provider-specific. I think it might be better if you surface it through a callback function that's specific to PayPal Commerce. i.e.:

```ts
initializePayment({
    paypalcommerce: {
        onRender() {
            // When this is called, the caller would know that PayPal button is rendered and
            // therefore can remove the standard submit button. And when the user switches 
            // to a different payment method, the caller can restore the standard submit button.
        }
    }
})
```

Reply:
Thank you for suggestions!

Yes, we should render embedded submit button only in case when shopper didn't approve PayPal account (i.e.: cart page). I think submit button depends on the checkout-sdk and in my case I call `enableEmbeddedSubmitButton ` [here](https://github.com/bigcommerce/checkout-sdk-js/pull/980/files#diff-3344e5b56907eaf188a8208fc12e5a51R62) if we should hide standard button. And call `disableEmbeddedSubmitButton` in the `deinitialize` [here](https://github.com/bigcommerce/checkout-sdk-js/pull/980/files#diff-3344e5b56907eaf188a8208fc12e5a51R105) for remove embedded button.
This actions change `isShowEmbeddedSubmitButton` in state and it use in checkout-js [in this PR](https://github.com/bigcommerce/checkout-js/pull/403). And in my point of view it work like `isInitializing`, `isSubmitting` and others ways of communication between two applications

I'll try your idea with `onRender`. But I have some questions. Do you mean call some action in `onRender` for changing state of `PaymentForm` and render standard or embedded button? Correct?

Reply:
@davidchin what are the possible downsides for having communication via `CheckoutState`? I understand the concern about the leakage of the provider-specific data, however, as Olga mentioned above, we think that `isShowEmbeddedSubmitButton` would be the default way of communication between two apps along with `isInitializing`, `isSubmitting`